Queue/Processor/Handler.php

```diff
@@ -67,6 +68,15 @@ public function onException(RequestSet $requestSet, Exception $e)
         // todo: how do we want to handle DbException or RedisException?
         $this->hasError = true;
 
+        Common::printDebug('Got exception: ' . $e->getMessage());
+
+        // retry if a deadlock or a lock wait timeout happened
+        if (Db::get()->isErrNo($e, 1213) || Db::get()->isErrNo($e, 1205)) {
+            $this->requestSetsToRetry[] = $requestSet;
```

Comment:
I haven't looked much in the PR yet but there might be a problem if a request set contains multiple requests and some have been recorded while some others not. I would probably only perform this logic if the request set contains only one request.

Reply:
In this case Handler->process()  would loop through the requests and the exception would come from this line https://github.com/matomo-org/plugin-QueuedTracking/blob/57c9b9710fae47f2773b2a0c21c933508df8da2b/Queue/Processor/Handler.php#L50.

Then in Processor->processRequestSets() we would catch it here https://github.com/matomo-org/plugin-QueuedTracking/blob/57c9b9710fae47f2773b2a0c21c933508df8da2b/Queue/Processor.php#L195.

And then, after looping through all requests sets, revert the whole transaction here https://github.com/matomo-org/plugin-QueuedTracking/blob/57c9b9710fae47f2773b2a0c21c933508df8da2b/Queue/Processor.php#L205.

Side note: what is a bit confusing here ,is that Handler->process() will on success add the complete reuquest set to requestSetsToRetry. Which is correct because we want to retry it, if an exception happens and requestSetsToRetry is emptied during Handler->commit()

Reply:
Does it maybe make sense to implement this in core eg in https://github.com/matomo-org/matomo/blob/3.10.0-b1/core/Tracker/Handler.php#L49 and reply the failed request if needed? Could also be useful for regular requests

Reply:
@MichaelHeerklotz just realising this probably doesn't make sense in core since we aren't using transactions there in the same way and we would be risking that maybe some parts would be executed twice. The only way would be wrap SQL inserts/updates in `Tracker\Model` https://github.com/matomo-org/matomo/blob/3.10.0-b1/core/Tracker/Model.php with the try / catch to ensure to only retry the specific query

Reply:
@tsteur Yes, without a transaction it gets complicated. I think some major changes would be needed to make it work.